ministryofjustice/correspondence_tool_staff
# [03] CT 2192 Add sync operations (PR No. 3)

## Description
Implements the synchronisation/reconciliation mechanism for CaseReport constituent data changes. 

## Self-review checklist

 
### Related JIRA tickets
https://dsdmoj.atlassian.net/browse/CT-2192
 
### Deployment
N/A
 
### Manual testing instructions
1. Edit any Case, User, Team and note `warehouse_case_reports` is consistent
2. Assigning a User should ensure the CaseReport is consistent
3. Closing a case shouold ensure `CaseReport` is consistent

### Notes
The implementation is fairly basic using the `after_commit` hook on `ApplicationRecord` to check whether the Model is:

1. One of the Models that contributes data to the 'Cases Report'
2. If so then the `sync` operation is queued as a Job. This is to offload the work to a separate thread/container and prevent the web application from hanging, and to reduce resource usage on the main server.
3. The sync operation may update multiple `CaseReport` entries depending on the type of data that is being updated e.g. a User's Name changing could affect multiple CaseReport entries
